### ISSUE TITLE
Vendored library linking failure

### ISSUE THREAD

**Issue by laramiel** [bazelbuild/rules_go#8]
When one vendored library depends on another, the current rules have a build failure.

**Repro:** https://github.com/laramiel/bazel-golang-vendor-test.git

* This repository includes https://github.com/tinylib/msgp and https://github.com/philhofer/fwd as vendored libraries.
* tinylib/msgp depends on philhofer/fwd
* Each library includes tests.

```sh
$ git clone https://github.com/laramiel/bazel-golang-vendor-test.git
$ cd bazel-golang-vendor-test
$ bazel test //vendor/...
```

**Expected:**  tests in all subdirectories pass.
**Actual:** tinylib/msgp tests fail to link.

When run with `bazel test //vendor/... --verbose_failures` it appears that the go tool link command is looking for the fwd library in the wrong location.

**Comment by kamalmarhubi** [comment]
This appears to be a show stopper for using bazel for go projects with vendored libraries. Could one of the developers comment?

**Comment by pnasrat** [comment]
Investigating

With raw repo

```
bazel test //vendor/github.com/philhofer:fwd_test
INFO: Found 1 test target...
Target //vendor/github.com/philhofer:fwd_test up-to-date:
  bazel-bin/vendor/github.com/philhofer/fwd_test.a
  bazel-bin/vendor/github.com/philhofer/fwd_test_main_test.go
  bazel-bin/vendor/github.com/philhofer/fwd_test_main_test.a
  bazel-bin/vendor/github.com/philhofer/fwd_test
INFO: Elapsed time: 0.421s, Critical Path: 0.18s
//vendor/github.com/philhofer:fwd_test                                   PASSED in 0.1s

Executed 1 out of 1 tests: 1 test passes.
There were tests whose specified size is too big. Use the --test_verbose_timeout_warnings command line option to see which ones these are.
```

```
bazel test --verbose_failures //vendor/github.com/tinylib/msgp:msgp_test
INFO: Found 1 test target...
ERROR: /home/pnasrat/.go/src/github.com/laramiel/bazel-golang-vendor-test/vendor/github.com/tinylib/msgp/BUILD:40:1: error executing shell command: 'rm -rf bazel-out/local-fastbuild/bin/vendor/github.com/tinylib/msgp/msgp_test.dir && mkdir -p bazel-out/local-fastbuild/bin/vendor/github.com/tinylib/msgp/msgp_test.dir && mkdir -p bazel-out/local-...' failed: namespace-sandbox failed: error executing command
  (cd /home/pnasrat/.cache/bazel/_bazel_pnasrat/f5af6ebd7ac9c80083f449463ddddecf/bazel-golang-vendor-test && \
  exec env - \
    GOARCH=amd64 \
    GOOS=linux \
  /home/pnasrat/.cache/bazel/_bazel_pnasrat/f5af6ebd7ac9c80083f449463ddddecf/bazel-golang-vendor-test/_bin/namespace-sandbox @/home/pnasrat/.cache/bazel/_bazel_pnasrat/f5af6ebd7ac9c80083f449463ddddecf/bazel-golang-vendor-test/bazel-sandbox/108cb469-94b8-48eb-9aa9-89502b7eb6e4-0.params -- /bin/bash -c 'rm -rf bazel-out/local-fastbuild/bin/vendor/github.com/tinylib/msgp/msgp_test.dir && mkdir -p bazel-out/local-fastbuild/bin/vendor/github.com/tinylib/msgp/msgp_test.dir && mkdir -p bazel-out/local-fastbuild/bin/vendor/github.com/tinylib/msgp/msgp_test.dir/github.com/laramiel/bazel-golang-vendor-test/vendor/github.com/tinylib/msgp && ln -s ../../../../../../../../../../../../../../../bazel-out/local-fastbuild/bin/vendor/github.com/tinylib/msgp/msgp_test bazel-out/local-fastbuild/bin/vendor/github.com/tinylib/msgp/msgp_test.dir/github.com/laramiel/bazel-golang-vendor-test/vendor/github.com/tinylib/msgp/msgp_test && mkdir -p bazel-out/local-fastbuild/bin/vendor/github.com/tinylib/msgp/msgp_test.dir/github.com/laramiel/bazel-golang-vendor-test/vendor/github.com/tinylib/msgp && ln -s ../../../../../../../../../../../../../../../bazel-out/local-fastbuild/bin/vendor/github.com/tinylib/msgp/msgp_test.a bazel-out/local-fastbuild/bin/vendor/github.com/tinylib/msgp/msgp_test.dir/github.com/laramiel/bazel-golang-vendor-test/vendor/github.com/tinylib/msgp/msgp_test.a && mkdir -p bazel-out/local-fastbuild/bin/vendor/github.com/tinylib/msgp/msgp_test.dir/github.com/laramiel/bazel-golang-vendor-test/vendor/github.com/tinylib/msgp && ln -s ../../../../../../../../../../../../../../../bazel-out/local-fastbuild/bin/vendor/github.com/tinylib/msgp/msgp_test_main_test.a bazel-out/local-fastbuild/bin/vendor/github.com/tinylib/msgp/msgp_test.dir/github.com/laramiel/bazel-golang-vendor-test/vendor/github.com/tinylib/msgp/msgp_test_main_test.a && export GOROOT=$(pwd)/external/golang_linux_amd64/go/bin/.. && cd bazel-out/local-fastbuild/bin/vendor/github.com/tinylib/msgp/msgp_test.dir && ../../../../../../../../external/golang_linux_amd64/go/bin/go tool link -L . -o github.com/laramiel/bazel-golang-vendor-test/vendor/github.com/tinylib/msgp/msgp_test github.com/laramiel/bazel-golang-vendor-test/vendor/github.com/tinylib/msgp/msgp_test_main_test.a').
/home/pnasrat/.cache/bazel/_bazel_pnasrat/f5af6ebd7ac9c80083f449463ddddecf/bazel-golang-vendor-test/external/golang_linux_amd64/go/pkg/tool/linux_amd64/link: cannot open file /home/pnasrat/.cache/bazel/_bazel_pnasrat/f5af6ebd7ac9c80083f449463ddddecf/bazel-golang-vendor-test/external/golang_linux_amd64/go/pkg/linux_amd64/github.com/laramiel/bazel-golang-vendor-test/vendor/github.com/philhofer/fwd.a: open /home/pnasrat/.cache/bazel/_bazel_pnasrat/f5af6ebd7ac9c80083f449463ddddecf/bazel-golang-vendor-test/external/golang_linux_amd64/go/pkg/linux_amd64/github.com/laramiel/bazel-golang-vendor-test/vendor/github.com/philhofer/fwd.a: no such file or directory
Target //vendor/github.com/tinylib/msgp:msgp_test failed to build
INFO: Elapsed time: 0.751s, Critical Path: 0.56s

Executed 0 out of 1 tests: 1 fails to build.
```

**Comment by pnasrat** [comment]
Using sandbox-debug and entering the sandbox you can see when we build the test we end up with

```
find bazel-out/
bazel-out/
bazel-out/local-fastbuild
bazel-out/local-fastbuild/bin
bazel-out/local-fastbuild/bin/vendor
bazel-out/local-fastbuild/bin/vendor/github.com
bazel-out/local-fastbuild/bin/vendor/github.com/tinylib
bazel-out/local-fastbuild/bin/vendor/github.com/tinylib/msgp
bazel-out/local-fastbuild/bin/vendor/github.com/tinylib/msgp/msgp_test
bazel-out/local-fastbuild/bin/vendor/github.com/tinylib/msgp/msgp_test.dir
bazel-out/local-fastbuild/bin/vendor/github.com/tinylib/msgp/msgp_test.dir/github.com
bazel-out/local-fastbuild/bin/vendor/github.com/tinylib/msgp/msgp_test.dir/github.com/laramiel
bazel-out/local-fastbuild/bin/vendor/github.com/tinylib/msgp/msgp_test.dir/github.com/laramiel/bazel-golang-vendor-test
bazel-out/local-fastbuild/bin/vendor/github.com/tinylib/msgp/msgp_test.dir/github.com/laramiel/bazel-golang-vendor-test/vendor
bazel-out/local-fastbuild/bin/vendor/github.com/tinylib/msgp/msgp_test.dir/github.com/laramiel/bazel-golang-vendor-test/vendor/github.com
bazel-out/local-fastbuild/bin/vendor/github.com/tinylib/msgp/msgp_test.dir/github.com/laramiel/bazel-golang-vendor-test/vendor/github.com/tinylib
bazel-out/local-fastbuild/bin/vendor/github.com/tinylib/msgp/msgp_test.dir/github.com/laramiel/bazel-golang-vendor-test/vendor/github.com/tinylib/msgp
bazel-out/local-fastbuild/bin/vendor/github.com/tinylib/msgp/msgp_test.dir/github.com/laramiel/bazel-golang-vendor-test/vendor/github.com/tinylib/msgp/msgp_test
bazel-out/local-fastbuild/bin/vendor/github.com/tinylib/msgp/msgp_test.dir/github.com/laramiel/bazel-golang-vendor-test/vendor/github.com/tinylib/msgp/msgp_test_main_test.a
bazel-out/local-fastbuild/bin/vendor/github.com/tinylib/msgp/msgp_test.dir/github.com/laramiel/bazel-golang-vendor-test/vendor/github.com/tinylib/msgp/msgp_test.a
bazel-out/local-fastbuild/bin/vendor/github.com/tinylib/msgp/msgp_test_main_test.a
bazel-out/local-fastbuild/bin/vendor/github.com/tinylib/msgp/msgp_test.a
bazel-out/host
bazel-out/host/bin
bazel-out/host/bin/external
bazel-out/host/bin/external/io_bazel_rules_go
bazel-out/host/bin/external/io_bazel_rules_go/go
bazel-out/host/bin/external/io_bazel_rules_go/go/tools
bazel-out/host/bin/external/io_bazel_rules_go/go/tools/generate_test_main
bazel-out/host/bin/external/io_bazel_rules_go/go/tools/generate_test_main.runfiles
bazel-out/host/bin/external/io_bazel_rules_go/go/tools/generate_test_main.runfiles/external
bazel-out/host/bin/external/io_bazel_rules_go/go/tools/generate_test_main.runfiles/external/io_bazel_rules_go
bazel-out/host/bin/external/io_bazel_rules_go/go/tools/generate_test_main.runfiles/external/io_bazel_rules_go/go
bazel-out/host/bin/external/io_bazel_rules_go/go/tools/generate_test_main.runfiles/external/io_bazel_rules_go/go/tools
bazel-out/host/bin/external/io_bazel_rules_go/go/tools/generate_test_main.runfiles/external/io_bazel_rules_go/go/tools/generate_test_main
bazel-out/host/bin/external/io_bazel_rules_go/go/tools/generate_test_main.a
```

**Comment by pnasrat** [comment]
If we do a clean and then try build `bazel build //vendor/github.com/tinylib/msgp:go_default_library`

Note this is using my fork at https://github.com/pnasrat/bazel-golang-vendor-test 

```
 find bazel-out/
bazel-out/
bazel-out/volatile-status.txt
bazel-out/local-fastbuild
bazel-out/local-fastbuild/bin
bazel-out/local-fastbuild/bin/vendor
bazel-out/local-fastbuild/bin/vendor/github.com
bazel-out/local-fastbuild/bin/vendor/github.com/tinylib
bazel-out/local-fastbuild/bin/vendor/github.com/tinylib/msgp
bazel-out/local-fastbuild/bin/vendor/github.com/tinylib/msgp/go_default_library.a
bazel-out/local-fastbuild/bin/vendor/github.com/philhofer
bazel-out/local-fastbuild/bin/vendor/github.com/philhofer/fwd
bazel-out/local-fastbuild/bin/vendor/github.com/philhofer/fwd/go_default_library.a
bazel-out/local-fastbuild/testlogs
bazel-out/local-fastbuild/genfiles
bazel-out/stable-status.txt
```

I'll take a look at the skylark rules to see why the rules don't handle vendor deps properly in this case

**Comment by pnasrat** [comment]
`bazel query 'deps(//vendor/github.com/tinylib/msgp:msgp_test )' --output graph`
![graph](https://cloud.githubusercontent.com/assets/13798/14835719/3ff99d7a-0bd9-11e6-959e-ea2c4cf5f252.png)

**Comment by pnasrat** [comment]
Additional diagnosis using `bazel query` 

```
 bazel query --output=build //vendor/github.com/tinylib/msgp:all
# /home/pnasrat/.go/src/github.com/laramiel/bazel-golang-vendor-test/vendor/github.com/tinylib/msgp/BUILD:42:1
go_test(
  name = "msgp_test",
  library = "//vendor/github.com/tinylib/msgp:go_default_library",
  srcs = ["//vendor/github.com/tinylib/msgp:msgp/edit_test.go", "//vendor/github.com/tinylib/msgp:msgp/extension_test.go", "//vendor/github.com/tinylib/msgp:msgp/floatbench_test.go", "//vendor/github.com/tinylib/msgp:msgp/json_bytes_test.go", "//vendor/github.com/tinylib/msgp:msgp/json_test.go", "//vendor/github.com/tinylib/msgp:msgp/number_test.go", "//vendor/github.com/tinylib/msgp:msgp/raw_test.go", "//vendor/github.com/tinylib/msgp:msgp/read_bytes_test.go", "//vendor/github.com/tinylib/msgp:msgp/read_test.go", "//vendor/github.com/tinylib/msgp:msgp/write_bytes_test.go", "//vendor/github.com/tinylib/msgp:msgp/write_test.go"],
)

# /home/pnasrat/.go/src/github.com/laramiel/bazel-golang-vendor-test/vendor/github.com/tinylib/msgp/BUILD:5:1
_go_prefix_rule(
  name = "go_prefix",
  visibility = ["//visibility:public"],
  generator_name = "go_prefix",
  generator_function = "go_prefix",
  generator_location = "vendor/github.com/tinylib/msgp/BUILD:5",
  prefix = "github.com/tinylib/msgp",
)

# /home/pnasrat/.go/src/github.com/laramiel/bazel-golang-vendor-test/vendor/github.com/tinylib/msgp/BUILD:7:1
go_library(
  name = "go_default_library",
  visibility = ["//visibility:public"],
  deps = ["//vendor/github.com/philhofer/fwd:go_default_library"],
  srcs = ["//vendor/github.com/tinylib/msgp:msgp/advise_linux.go", "//vendor/github.com/tinylib/msgp:msgp/circular.go", "//vendor/github.com/tinylib/msgp:msgp/defs.go", "//vendor/github.com/tinylib/msgp:msgp/edit.go", "//vendor/github.com/tinylib/msgp:msgp/elsize.go", "//vendor/github.com/tinylib/msgp:msgp/errors.go", "//vendor/github.com/tinylib/msgp:msgp/extension.go", "//vendor/github.com/tinylib/msgp:msgp/file.go", "//vendor/github.com/tinylib/msgp:msgp/integers.go", "//vendor/github.com/tinylib/msgp:msgp/json.go", "//vendor/github.com/tinylib/msgp:msgp/json_bytes.go", "//vendor/github.com/tinylib/msgp:msgp/number.go", "//vendor/github.com/tinylib/msgp:msgp/read.go", "//vendor/github.com/tinylib/msgp:msgp/read_bytes.go", "//vendor/github.com/tinylib/msgp:msgp/size.go", "//vendor/github.com/tinylib/msgp:msgp/unsafe.go", "//vendor/github.com/tinylib/msgp:msgp/write.go", "//vendor/github.com/tinylib/msgp:msgp/write_bytes.go"],
)
```

**Comment by pnasrat** [comment]
Used `-s` to look at the actions so the initial looks we correctly identify the deps 

```
>>>>>>>>> # //vendor/github.com/tinylib/msgp:msgp_test [action 'Creating runfiles tree bazel-out/local-fastbuild/bin/vendor/github.com/tinylib/msgp/msgp_test.runfiles']
>>>>>>>>> # @io_bazel_rules_go//go/tools:generate_test_main [action 'Creating runfiles tree bazel-out/host/bin/external/io_bazel_rules_go/go/tools/generate_test_main.runfiles [for host]']
>>>>>>>>> # @io_bazel_rules_go//go/tools:generate_test_main [action 'GoCompile external/io_bazel_rules_go/go/tools/generate_test_main.a [for host]']
>>>>>>>>> # //vendor/github.com/philhofer/fwd:go_default_library [action 'GoCompile vendor/github.com/philhofer/fwd/go_default_library.a']
>>>>>>>>> # @io_bazel_rules_go//go/tools:generate_test_main [action 'GoLink external/io_bazel_rules_go/go/tools/generate_test_main [for host]']
>>>>>>>>> # //vendor/github.com/tinylib/msgp:msgp_test [action 'GoCompile vendor/github.com/tinylib/msgp/msgp_test.a']
>>>>>>>>> # //vendor/github.com/tinylib/msgp:msgp_test [action 'GoTestGenTest vendor/github.com/tinylib/msgp/msgp_test_main_test.go']
>>>>>>>>> # //vendor/github.com/tinylib/msgp:msgp_test [action 'GoCompile vendor/github.com/tinylib/msgp/msgp_test_main_test.a']
>>>>>>>>> # //vendor/github.com/tinylib/msgp:msgp_test [action 'GoLink vendor/github.com/tinylib/msgp/msgp_test']
```

**Comment by pnasrat** [comment]
I suspect this is a duplicate of Issue #6

**Comment by Limdi** [comment]
Possibly, running the tests with my workaround makes them pass.

**Comment by yugui** [comment]
Fixed with #33.
